### PR TITLE
CMake Implementation Selection

### PR DESCRIPTION
**Pull request by LeStarch**
| | |
|:---|:---|
|**_Originating Project/Creator_**| |
|**_Affected Component_**|  |
|**_Affected Architectures(s)_**|  |
|**_Related Issue(s)_**|  |
|**_Has Unit Tests (y/n)_**|  |
|**_Builds Without Errors (y/n)_**|  |
|**_Unit Tests Pass (y/n)_**|  |
|**_Documentation Included (y/n)_**|  |

---
## Change Description

This adds in the CMake functionality to allow choices of various implementation sets backed by CMake.  The canonical example is choosing task implementations for the OS layer.

These changes were previously reviewed here: https://github.com/nasa/fprime/pull/2121

I've decided that a long-running feature branch is not the way to go due to the difficulty of keeping up with `nasa/devel`.  It is better to refactor individual OSAL units one at a time, leaving the others.

## Rationale

Keeping up with nasa/devel is challenging.  Better to reduce the PR to unit-steps.

## Testing/Review Recommendations

Ensure stock builds work.

## Future Work

Refactoring each OSAL subunit one at a time.

### CODE REVIEW

File: docs/Design/package-implementations.md
Comment (thomas-bc) [inline docs/Design/package-implementations.md:1]: Links to this page should probably be added to:
- TOC: https://github.com/nasa/fprime/blob/devel/docs/_includes/toc.md
- Design index: https://github.com/nasa/fprime/blob/devel/docs/Design/general.md

File: docs/Design/package-implementations.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+## Choosing an Implementation
+
+Platform developers *must* specify an implementation of every package used in the system. Failing to do so means that a given functionality is undefined and impossible to link.  Stub implementations are provided in the case that a platform does not support a given package's functionality.
+
+Choosing an implementation is done with a series of `choose_fprime_implementation` calls. One for each required implementation defined in the system.  These calls are put in a platform CMake file.
+
+## Overriding an Implementation Choice
+
+Some executables, unit tests, and deployments may wish to use a different implementation than that specified by the platform. This can be done by a `choose_fprime_implementation` call in the deployment, executable, or unit test's module. For example, a unit test may which to choose `Os_File_Stub` as an implementation of `Os_File` to disable platform file system support for the given unit test.
```
Comment (thomas-bc) [inline docs/Design/package-implementations.md]: Typo: "may which" --> "may wish" ?

File: docs/Design/package-implementations.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# Package Implementations
+
+Certain parts of F´ have different implementations that a project may choose from.  The canonical example of this is a file system implementation. Projects may wish to use the stubbed implementation of the file system, the platform supplied standard implementation, or something project-specific like a flash chip or SD card.
+
+These packages are must be fully specified. Thus, every platform must specify a specific implementation for every package. These specific implementations may be overridden for a specific executable deployment or unit test.
```
Comment (thomas-bc) [inline docs/Design/package-implementations.md]: Typo: "These packages *are* must be ..."

**Review by thomas-bc — APPROVED** [review]
LGTM! Couple of comments with the doc

File: docs/Design/package-implementations.md
Comment (LeStarch) [reply docs/Design/package-implementations.md:1]: Agreed.

File: docs/Design/package-implementations.md (hunk 1 above)
Comment (LeStarch) [reply docs/Design/package-implementations.md]: yes!

File: docs/Design/package-implementations.md (hunk 2 above)
Comment (LeStarch) [reply docs/Design/package-implementations.md]: Agreed!

**Comment by LeStarch** [comment]
Fixed!

**Review by thomas-bc — APPROVED** [review]
Ready to merge whenever you are 🚢

**Pull request merged**